File: geocomp/convhull/convhull.cpp
#include <bits/stdc++.h>
using namespace std;

ifstream in("infasuratoare.in");
ofstream out("infasuratoare.out");

struct Point {
    double x, y;

    bool operator<(Point rhs) const {
        return make_tuple(x, y) < make_tuple(rhs.x, rhs.y);
    }

    friend istream& operator>>(istream& is, Point& pt) {
        return is >> pt.x >> pt.y;
    }

    friend ostream& operator<<(ostream& os, const Point& pt) {
        return os << pt.x << ' ' << pt.y;
    }

    Point operator-(Point b) const {
        return { x - b.x, y - b.y };
    }
};

double cross_product(Point u, Point v) {
    return u.x * v.y - u.y * v.x;
}

double orientation_test(Point a, Point b, Point c) {
    return cross_product(b - a, c - a);
}

bool right_turn(Point a, Point b, Point c) {
    return orientation_test(a, b, c) < 0;
}

vector<Point> compute_hull(const vector<Point>& points) {
    vector<Point> hull = { points[0], points[1] };

    int i = 2;
    while (i < points.size()) {
        hull.push_back(points[i]);
        while (hull.size() >= 3
            && !right_turn(hull.rbegin()[2], hull.rbegin()[1], hull.rbegin()[0]))
        {
            hull.erase(hull.begin() + hull.size() - 2);
        }
        ++i;
    }

    return hull;
}

template <typename T>
void print_vector(const vector<T>& v) {
    for (auto elem : v) {
        out << elem << '\n';
    }
}

int main() {
    int n;
    in >> n;

    vector<Point> points(n);
    for (int i = 0; i < n; ++i) {
        in >> points[i];
    }

    sort(points.begin(), points.end());

    auto lower_hull = compute_hull(points);
    lower_hull.erase(lower_hull.begin());

    reverse(points.begin(), points.end());

    auto upper_hull = compute_hull(points);
    upper_hull.erase(upper_hull.begin());

    vector<Point> hull = lower_hull;
    hull.insert(hull.end(), upper_hull.cbegin(), upper_hull.cend());

    out << hull.size() << '\n';
    print_vector(hull);
}

File: geocomp/convhull/convhull.py
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    @staticmethod
    def read(f):
        x, y = map(float, next(f).split())
        return Point(x, y)

    def __lt__(self, other):
        return (self.x, self.y) < (other.x, other.y)

    def __sub__(self, other):
        return Point(self.x - other.x, self.y - other.y)

    def __repr__(self):
        return '{} {}'.format(self.x, self.y)


def cross_product(u, v):
    return u.x * v.y - u.y * v.x


def orientation_test(a, b, c):
    return cross_product(b - a, c - a)


def right_turn(a, b, c):
    return orientation_test(a, b, c) < 0


def compute_hull():
    hull = [points[0], points[1]]

    i = 2
    while i < len(points):
        hull.append(points[i])
        while len(hull) >= 3 and not right_turn(hull[-3], hull[-2], hull[-1]):
            hull.pop(-2)
        i += 1

    return hull


fin = open('infasuratoare.in')

n = int(next(fin))

points = [Point.read(fin) for _ in range(n)]

points.sort()

lower_hull = compute_hull()

points.reverse()
upper_hull = compute_hull()

hull = lower_hull[1:] + upper_hull[1:]

with open('infasuratoare.out', 'w') as fout:
    print(len(hull), *hull, sep='\n', file=fout)
